torch/_dynamo/guards.py

```diff
@@ -511,6 +521,106 @@ def combine_scopes(left, right):
         self.check_fn = self.compile_check_fn(local_builder, global_builder)
         self._seen_ids.clear()
 
+    """
+    This is a complex bit of logic. The outline here is brief. For a line by line breakdown, see
+    the code comments below.
+
+    The role of this function is to take the current state of symbolic shape guards, tensor ids in the
+    CURRENT dynamo frame, and tensor names (dynamo's frame agnostic tensor reference mechanism, see TensorCheck and
+    guards.cpp for more info) - and produce executable python expressions for addition to our guarded code components
+    that make their way into check_fn.
+
+    We DO NOT create guards based on ids. The IDs act as a lookup for the following mapping:
+
+    dynamo: tensor_name <> tensor_id
+    shape_env: tensor_id <> shape_expr
+
+    This allows us to then create a tensor_name <> shape_expr association for the current frames guards.
```

Comment:
How come the shape env can't directly store tensor name instead

Reply:
I dont think it has that data readily available - and plumbing names down to shape_env is a lot of grossness. It's easier to bring guards up, imo.

Reply:
Actually, I take it back. All of this compute happens while the Tensors are guaranteed to stay live (because we haven't actually run the frame yet, right?) So just use Tensors directly, don't bother with id.

Reply:
id is more manageable, we don't really need access to the tensor itself at this point, right?

Reply:
Tensor is more debuggable though, because the id() can be subject to ABA problems with dealloc-realloc; also you can actually poke the tensor in debugging to see which one it is.

Reply:
Discussed offline, keeping it as id, fixing lifecycle by lifting this to a purely dynamo level.